### ISSUE TITLE
Modifying the request headers doesn't work

### ISSUE THREAD

**Issue by ARochniak** [amannn/next-intl#266]
### Description

Using `next-intl@2.14.0-beta.3` version and following the example [withExtraMiddleware](https://github.com/amannn/next-intl/issues/247#issuecomment-1523292342) the request headers aren't modified.

### Mandatory reproduction URL (CodeSandbox or GitHub repository)

https://github.com/ARochniak/next-intl-middleware.git

### Reproduction description

Steps to reproduce:
1. Clone reproduction repo
2. Run `yarn && yarn dev`
3. Open `https://localhost:3000` in the browser
4. Restart the page
5. You can see in the log message `X-test header: null`, but it should be `X-test header: test`

### Expected behaviour

According to the example [withExtraMiddleware](https://github.com/amannn/next-intl/issues/247#issuecomment-1523292342), request headers should be modified correctly.

**Comment by ARochniak** [comment]
Hello @amannn,
Firstly, thanks for your work 👍
As for the issue, it can be fixed by modifying the middleware's `getResponseInit` function like that:
``` typescript
function getResponseInit() {
  if (hasOutdatedCookie) {
    // Only apply a header if absolutely necessary
    // as this causes full page reloads
    request.headers.set(HEADER_LOCALE_NAME, locale);
  }
  const responseInit = {
    request: {
      headers: request.headers
    }
  };

  return responseInit;
}
```
I can create a PR for this if it's an appropriate solution.

**Comment by amannn** [comment]
Hello @ARochniak and thanks for the bug report!

You're right, seems like I've overlooked this. If you're interested to create a PR, that would be fantastic!

I noticed that the bug is already present in the stable release. Would you like to create a PR that targets the main branch with a fix and [test](https://github.com/amannn/next-intl/blob/main/packages/next-intl/test/middleware/middleware.test.tsx) for this? From there, we can sync it to the RSC branch afterward.

When syncing to the RSC branch, we need to be careful that we don't trigger more page reloads than necessary (see the comment in your code snippet).

**Comment by ARochniak** [comment]
@amannn Ok, I will process it later today.
As for the page reloads issue, I guess it only relates to the earlier versions of next 13, because I also faced with this issue while was testing the `middleware` in the past. But for now it works fine for me, without full page reloads. Maybe it's necessary to highlight it somewhere (at least left a comment), that it won't work with the earliest versions of next13.

**Comment by amannn** [comment]
Cool, thanks a lot!

> As for the page reloads issue, I guess it only relates to the earlier versions of next 13, because I also faced with this issue while was testing the `middleware` in the past. But for now it works fine for me, without full page reloads.

That would be great, I haven't tried it in some time.

Let me know if I can help with something!

**Comment by ARochniak** [comment]
@amannn FYI, the `full page reload` issue was fixed in the `next 13.2.0`.
